ООП ч2 Полиморфизм и интерфейсы.

Boroda9992 opened the pull request:
Использовал абстрактный класс-предок Vehicle для наследования общих свойств. 
В интерфейсе объявил дефолтную реализацию всех методов нашей сервисной станции. 
В ServiceStation подаю объект Vehicle и определяю по количеству колес что за транспорт и какие к нему применять методы. Для нашего примера такого сравнения вполне достаточно - транспортные средства вполне можно разнести по типам от количества колес и для каждого есть название (если не ошибаюсь - для грузовиков уже считается количество осей).  Так или иначе, теперь не надо думать куда вводить null, чтобы программа сработала как надо.

Boroda9992 committed:
ООП ч2 Полиморфизм и интерфейсы.

Boroda9992 committed:
ООП ч2 Полиморфизм и интерфейсы. Исправления 1.
Убрал дефолтную реализацию методов в интерфейсе VehicleService. Был уверен, что обязан хоть какую-то реализацию задать, раз уж использовал default.
Оставил один интерфейс VehicleService. В коде есть комментарий, почему я не сделал три разных.
Убрал проверку по числу колес. Я считал эту проверку частью задания - в коде должен быть механизм выбора транспортного средства. Число колес я нашел подходящим, потому что обычно автотранспорт вполне возможно классифицировать по числу колес и осей. Так или иначе, без этого if else стало проще, но если этьо часть задания - пожалуй сделал бы через новый параметр, например строку и проверял бы ее у транспортного средства, чтобы отнести его к конкретному типу (и, в идеале применить, свой отдельный интерфейс).

dmitry-bizin commented on .idea/.gitignore:
папку .idea надо игнорировать, а не коммитить в репозиторий

Boroda9992 replied:
В gitignore у меня все было как обычно, перед первым коммитом я прописал что не надо добавлять в репозиторий - как обычно...
В любом случае, я уже разобрался как убрать эти файлы из очереди через консоль.

dmitry-bizin replied:
Можно не через консоль, а просто перед коммитом в идее галки убрать с мусорных файлов, я про это говорил на самом первом уроке.

Boroda9992 replied:
Я именно так и сделал, но все равно ушло в коммит. Есть идеи почему?

dmitry-bizin replied:
Ну если бы так и сделал, то ничего бы не ушло в коммит, потому что это именно те файлы, который выбираешь для коммита. Видимо, невнимательно выбрал или забыл выбрать, как обычно бывает, и закоммитил всё. Если что, ниже скрин про чекбоксы при коммите, я про это
![image](https://github.com/Boroda9992/OOP_pt2_polymorphysm_and_interfaces/assets/15345076/967cca5f-f056-40ce-ba36-b845be6c82d8)

dmitry-bizin commented on src/Bicycle.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public void updateTyres() {
+        System.out.println("[" + getModelName() + "] " + getWheelsCount() + " Tyres updated!");
```
чтобы был явно виден полиморфизм реализация метода должна отличаться в наследниках, лучше всего просто написать, что идёт проверка колёс именно у велосипеда

Boroda9992 replied:
В задании изначально был цикл, который считал количество колес и каждый раз писал, что покрышка заменена. Я решил, что должен отражать количество колес в выводе (каждое колесо стоит денег, как минимум) так или иначе. Изначально я пытался именно делать приписку по названию класса - "_Bicycle_ tyre updated!", однако для этого я собирался добавлять параметр, в нем хранить строку и эту строку, передавать ее в метод... Мне показалось, что горожу лишнего и я решил, что лучшим решением будет использовать то, что у нас уже есть в коде, если можно не писать больше. Тем более, что мы явно указываем класс в модели, а в реальной жизни мы вполне возможно сможем понять по модели к каком классу относится тот или иной транспорт (раз уж занимаемся заменой покрышек)

dmitry-bizin replied:
> Я решил, что должен отражать количество колес в выводе

Ну цикл это как раз и делал - выводил, что проверил такое-то колесо. Сколько раз вывел, столько и колёс, вроде информативнее некуда)

> однако для этого я собирался добавлять параметр, в нем хранить строку и эту строку, передавать ее в метод

Зачем? Не нужно никаких параметров, суть полиморфизма как раз в том, что сам класс меняет поведение, т.е. без всяких параметров надо просто в классе велосипеда было написать, что идёт проверка колёс именно велосипеда, а не каких-то абстрактных колёс.

dmitry-bizin commented on src/Car.java:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    @Override
+    public void updateTyres() {System.out.println("[" + getModelName() + "] " + getWheelsCount() + " Tyres updated!");}
```
чтобы был явно виден полиморфизм реализация метода должна отличаться в наследниках, лучше всего просто написать, что идёт проверка колёс именно у машины

dmitry-bizin commented on src/Car.java:
```diff
@@ -0,0 +1,17 @@
+    public void updateTyres() {System.out.println("[" + getModelName() + "] " + getWheelsCount() + " Tyres updated!");}
+
+    @Override
+    public void checkEngine() {System.out.println("[" + getModelName() + "] Engine checked!");}
```
чтобы был явно виден полиморфизм реализация метода должна отличаться в наследниках, лучше всего просто написать, что идёт проверка двигателя именно у машины

dmitry-bizin commented on src/ServiceStation.java:
```diff
@@ -0,0 +1,45 @@
+     * Со временем, это приведет к тому, что у нас будет один метод-свалка check, перебирающий все доступные нам операции над транспортным средством, только ради того, чтобы выполнить одну-две.
+     * @param vehicle - абстрактный родительский класс, присущий любому транспортному средству, с которым работает программа.
+     */
+    public void check(Vehicle vehicle) {
+        vehicle.updateTyres();
+        vehicle.checkEngine();
+        vehicle.checkTrailer();
+    }
+
+    /**
+     * Специализированный метод для обслуживания велосипедов.
+     * Может быть полезен, если у нас появятся разные типы велосипедов, скидка для велосипедистов, появится необходимость выгружать бухгалтерскую отчетность по видам транспорта (потому что там налог, нарпимер, разный) и т.д.
+     * Перебирает все операции, которые мы можем проводить над велосипедом.
+     * @param bicycle - объект класса Bicycle, потомка Vehicle
+     */
+    public void checkBicycle (Bicycle bicycle){
+        bicycle.updateTyres();
+    }
+
+    /**
+     * Специализированный метод для обслуживания седанов.
+     * Перебирает все операции, которые мы можем проводить над седаном.
+     * @param car - объект класса Car, потомка Vehicle
+     */
+    public void checkCar (Car car){
+        car.updateTyres();
+        car.checkEngine();
+    }
+
+    /**
+     * Специализированный метод для обслуживания пикапов.
+     * Перебирает все операции, которые мы можем проводить над пикапом.
+     * @param truck - объект класса Truck, потомка Vehicle
+     */
+    public void checkTruck (Truck truck){
+        truck.updateTyres();
+        truck.checkEngine();
+        truck.checkTrailer();
+    }
```
как я и говорил на разборе такой способ решения не укладывается в тему домашки - ООП, т.к. у тебя нет полиморфизма ООП, а есть простая перегрузка, которая не имеет никакого отношения к ООП. Настоятельно рекомендую сделать именно через полиморфизм ООП, чтобы понять его суть

dmitry-bizin commented on src/Truck.java:
```diff
@@ -0,0 +1,20 @@
+        super(modelName, wheelsCount);
+    }
+    @Override
+    public void updateTyres() {System.out.println("[" + getModelName() + "] " + getWheelsCount() + " Tyres updated!");}
```
чтобы был явно виден полиморфизм реализация метода должна отличаться в наследниках, лучше всего просто написать, что идёт проверка колёс именно у грузовика

dmitry-bizin commented on src/Truck.java:
```diff
@@ -0,0 +1,20 @@
+    public void updateTyres() {System.out.println("[" + getModelName() + "] " + getWheelsCount() + " Tyres updated!");}
+
+    @Override
+    public void checkEngine() {System.out.println("[" + getModelName() + "] Engine checked!");}
```
чтобы был явно виден полиморфизм реализация метода должна отличаться в наследниках, лучше всего просто написать, что идёт проверка двигателя именно у грузовика

dmitry-bizin commented on src/VehicleService.java:
```diff
@@ -0,0 +1,16 @@
+    //В данной конкретной задаче получится больше кода, если сделать все правильно, чем если оставить так как есть. Я не
+    // сделал этого не потому, что не знаю, а потому что не хочу городить три интерфейса по одной строчке =)
+
+    default void updateTyres (){}
+    default void checkEngine (){}
+    default void checkTrailer(){}
```
да, это решение неверное, т.к. таким образом у тебя у велосипеда появляется двигатель и прицеп - следствие неправильной проработки архитектуры классов. Здесь 2 пути решения проблемы:
1. Разбить этот интерфейс на 3 - в каждом по 1 методу. Это способ самый правильный. Подробнее можешь прочитать про принцип разделения интерфейсов в SOLID, это буква I там.
2. В этом интерфейсе сделать один метод всего, который будет называться check. Реализация этого метода и будет как раз разной в разных классах - велосипед проверяет только колёса, машина ещё и двигатель, а грузовик ещё и прицеп. Этот способ быстрее и попроще.

dmitry-bizin commented on src/VehicleService.java:
```diff
@@ -0,0 +1,16 @@
+    // в себя проверку давления в шинах, проверку протектора, проверку спиц, в то время, как проверка колес седана и пикапа
+    // не предполагает проверки спиц, да и вообще, использует другие процессы и инструменты.
+    //В данной конкретной задаче получится больше кода, если сделать все правильно, чем если оставить так как есть. Я не
+    // сделал этого не потому, что не знаю, а потому что не хочу городить три интерфейса по одной строчке =)
```
надо всегда писать код правильно, а не лишь бы сделать и сдать) иначе это (писать абы как) может войти в привычку, а плохие программисты мало кого интересуют)

Boroda9992 committed:
ООП ч2 Полиморфизм и интерфейсы. Исправления 2.
- Разбил общий интерфейс на отдельные, подключил к каждому классу транспорта свой набор интерфейсов.
- Реализовал класс ServiceStation, в нем - метод check() двумя разными способами - через if-else и через switсh - оба проверял, работают. Последний способ мне кажется более поддерживаемым. Так же, не уверен, что в реализации с if-else правильно использую getClass(). Как я понял, я могу сравнивать классы объектов через ==, потому что это не сами классы, а строки с их названием и нам не нужно здесь использовать equals().
- Внутри классов, в реализации методов, явно указал класс транспортного средства.

dmitry-bizin commented on src/ServiceEngine.java:
```diff
@@ -0,0 +1,4 @@
+public interface ServiceEngine {
+    default void checkEngine (){}
```
зачем делать дефолтный метод с пустым телом? какой в этом смысл? лучше сделать абстрактным и без тела, чем по дефолту ничего не делающим - абстрактным методам уделишь внимание, т.е. без их определения не скомпилируется код, а вот про пустой дефолтный метод можно забыть и потом долго не понимать, почему работает не так, как ожидается

dmitry-bizin commented on src/ServiceStation.java:
```diff
@@ -1,45 +1,43 @@
+ * Класс сервисной станции - для методов, применяемых непосредственно к объектам-наследникам Vehicle.
+ * Реализует: ServiceTrailer, ServiceEngine, ServiceTyres
+ */
+public class ServiceStation implements ServiceTyres, ServiceEngine, ServiceTrailer {
```
этот класс не должен реализовывать никакие интерфейсы, интерфейсы должны реализовывать только классы ТС

dmitry-bizin commented on src/ServiceStation.java:
```diff
@@ -1,45 +1,43 @@
-    public void checkBicycle (Bicycle bicycle){
-        bicycle.updateTyres();
-    }
+        String vehicleClassName = String.valueOf(vehicle.getClass());
+        switch (vehicleClassName){
+            case "class Bicycle":
+                vehicle.updateTyres();
+                break;
+            case "class Car":
+                vehicle.updateTyres();
+                vehicle.checkEngine();
+                break;
+            case "class Truck":
+                vehicle.updateTyres();
+                vehicle.checkEngine();
+                vehicle.checkTrailer();
+                break;
+        }
```
это всё неверно, это как раза обход полиморфизм. В базовом классе Vehicle надо добавить абстрактный метод проверки, который надо реализовать по-разному в классах-наследниках. Здесь же, в этом методе сервисной станции, будет всего 1 строчка кода - вызов этого абстрактного метода. Это и будет полиморфизм, сейчас у тебя его по-прежнему нет

dmitry-bizin commented on src/Bicycle.java:
```diff
@@ -1,15 +1,18 @@
+ * Перегружает: updateTyres()
  */
-public class Bicycle extends Vehicle implements VehicleService {
+public class Bicycle extends Vehicle implements ServiceTyres {
```
класс не должен реализовывать интерфейс, потому что базовый класс Vehicle уже это делает, дублировать не нужно

dmitry-bizin commented on src/Car.java:
```diff
@@ -1,17 +1,21 @@
+ * Перегружает: checkEngine(), updateTyres()
  */
-public class Car extends Vehicle implements VehicleService {
+public class Car extends Vehicle implements ServiceEngine, ServiceTyres {
```
класс не должен реализовывать интерфейсы, потому что базовый класс Vehicle уже это делает, дублировать не нужно

dmitry-bizin commented on src/Truck.java:
```diff
@@ -1,20 +1,26 @@
+ * Перегружает: checkTrailer(), checkEngine(), updateTyres()
  */
-public class Truck extends Vehicle implements VehicleService{
+public class Truck extends Vehicle implements ServiceTrailer, ServiceEngine, ServiceTyres{
```
класс не должен реализовывать интерфейсы, потому что базовый класс Vehicle уже это делает, дублировать не нужно

dmitry-bizin commented on src/ServiceTrailer.java:
```diff
@@ -0,0 +1,4 @@
+public interface ServiceTrailer {
+    default void checkTrailer(){}
```
зачем делать дефолтный метод с пустым телом? какой в этом смысл? лучше сделать абстрактным и без тела, чем по дефолту ничего не делающим - абстрактным методам уделишь внимание, т.е. без их определения не скомпилируется код, а вот про пустой дефолтный метод можно забыть и потом долго не понимать, почему работает не так, как ожидается

dmitry-bizin commented on src/ServiceTyres.java:
```diff
@@ -0,0 +1,3 @@
+public interface ServiceTyres {
+    default void updateTyres (){}
```
зачем делать дефолтный метод с пустым телом? какой в этом смысл? лучше сделать абстрактным и без тела, чем по дефолту ничего не делающим - абстрактным методам уделишь внимание, т.е. без их определения не скомпилируется код, а вот про пустой дефолтный метод можно забыть и потом долго не понимать, почему работает не так, как ожидается

Boroda9992 committed:
ООП ч2 Полиморфизм и интерфейсы. Исправления 3.
- В интерфейсах оставил только абстрактные методы. Изначально, по какой-то причине они у меня не работали и я использовал дефолтные методы без поведения, отметив, что надо повторить кое-что из пройденного. Повторить - повторил, а в чем была ошибка так и не понял. Может, дело в том, что я просто со свежей головой за проект сел.
- Логика работы методов сконцентрирована в абстрактном классе-родителе.
- В классах-наследниках переопределен check(), чтобы указать только те методы, которые применимы к конкретному классу-наследнику. Переопределен toString(), чтобы убрать фигурные скобки.
- в классе ServiceStation, по сути, вызывается только метод check(). Остальное - обертка для читаемости вывода.
